hydromt/gis/raster.py

```diff
@@ -981,6 +986,190 @@ def gen_zonal_stat(ds, geoms, stats, all_touched=False):
+            else:
+                p = geom.representative_point()
+
+            col, row = ~ds.raster.transform * (p.x, p.y)
+
+            col = int(np.floor(col))
+            row = int(np.floor(row))
+
+            if row < 0 or row >= ds.raster.height or col < 0 or col >= ds.raster.width:
+                logger.warning(
+                    f"Geometry at index {i} with {p} outside raster domain. Skipping."
+                )
+                continue
+
+            rows.append(row)
+            cols.append(col)
+            indices.append(i)
+
+        if not indices:
+            raise ValueError("All geometries outside raster domain")
+
+        # get sampled values
+        rows = np.asarray(rows)
+        cols = np.asarray(cols)
+
+        out = {}
+        bad_mask = np.zeros(len(indices), dtype=bool)
+
+        if variables is None:
+            variables = ds.data_vars.keys()
+
+        def _to_index_1d(_in: np.ndarray) -> tuple[str, np.ndarray]:
+            arr = np.squeeze(np.asarray(_in))  # shape (n_features, n_geoms)
+            if arr.ndim == 0:
+                arr = np.array([arr])  # ensure 1D
+            return ("index", arr)
+
+        # sample variables directly
+        for var in variables:
+            da = ds[var]
+            data = da.data
+
+            # Create array of sampled values (n_geoms, n_features)
+            sampled_vals = np.stack(
+                [
+                    data[..., r, c].compute()  # edge case for dask arrays
+                    if hasattr(data, "compute")
+                    else data[..., r, c]
+                    for r, c in zip(rows, cols, strict=True)
+                ],
+                axis=-1,
+            )
+            # Ensure shape is (n_features, n_geoms)
+            if sampled_vals.ndim == 1:
+                sampled_vals = sampled_vals[np.newaxis, :]
+
+            # check for nodata / nan values
+            nodata = da.raster.nodata
+            if nodata is not None:
+                bad_mask |= np.any(sampled_vals == nodata, axis=0)
+            bad_mask |= np.any(np.isnan(sampled_vals), axis=0)
+            if bad_mask.any():
+                exec_nodata_strat(
+                    f"{int(bad_mask.sum())} out of {len(indices)} geometries sampled nodata values for variable '{var}'.",
+                    handle_nodata,
+                )
+
+            # Determine stats
+            x_dim = ds.raster.x_dim
+            y_dim = ds.raster.y_dim
+            reduce_dims = [d for d in da.dims if d not in (x_dim, y_dim)]
+
+            if not reduce_dims:  # single band raster, no stats needed
+                if stats is not None:  # no stats requested
+                    logger.warning(
+                        "Statistics requested for single-band raster; returning sampled value."
+                    )
+                out[var] = _to_index_1d(sampled_vals)
+            else:  # compute stats
+                stats_to_apply = stats or ["mean"]
```

Comment:
I would just use "mean" as default argument. No need for None as possible type then and this would be more transpararent.